evm/db/journal.py

```diff
@@ -1,226 +1,264 @@
+                continue
+            else:
+                if value is None:
+                    raise KeyError(key)
+                else:
+                    return value
+        else:
+            raise KeyError(key)
+
+    def set(self, key, value):
+        self.latest[key] = value
+
+    def exists(self, key):
+        for changeset_data in reversed(self.journal_data.values()):
+            try:
+                value = changeset_data[key]
+            except KeyError:
+                continue
+            else:
+                if value is None:
+                    return False
+                else:
+                    return True
+        else:
+            return False
 
-    def __contains__(self, value):
-        return value in self.journal_data
+    def delete(self, key):
+        self.latest[key] = None
 
 
 class JournalDB(BaseDB):
     """
     A wrapper around the basic DB objects that keeps a journal of all changes.
-    Each time a snapshot is taken, the underlying journal creates a new
-    checkpoint.  The journal then keeps track of the original value for any
-    keys changed.  Reverting to a checkpoint involves merging the original key
-    data from any subsequent checkpoints into the given checkpoint giving
-    precidence earlier checkpoints.  Then the keys from this merged data set
-    are reset to their original values.
+    Each time a recording is started, the underlying journal creates a new
+    changeset and assigns an id to it. The journal then keeps track of all changes
+    that go into this changeset.
+
+    Discarding a changeset simply throws it away inculding all subsequent changesets
+    that may have followed. Commiting a changeset merges the given changeset and all
+    subsequent changesets into the previous changeset giving precidence to later
+    changesets in case of conflicting keys.
+
+    Nothing is written to the underlying db until `persist()` is called.
 
     The added memory footprint for a JournalDB is one key/value stored per
     database key which is changed.  Subsequent changes to the same key within
-    the same checkpoint will not increase the journal size since we only need
-    to track the original value for any given key within any given checkpoint.
+    the same changeset will not increase the journal size since we only need
+    to track latest value for any given key within any given changeset.
     """
     wrapped_db = None
-    journal = None
+    journal = None  # type: Journal
 
-    def __init__(self, wrapped_db):
+    def __init__(self, wrapped_db: BaseDB) -> None:
         self.wrapped_db = wrapped_db
-        self.journal = Journal()
+        self.reset()
 
-    def get(self, key):
-        return self.wrapped_db.get(key)
+    # TODO: Measure perf impact of exception ping pong
+    def get(self, key: bytes) -> bytes:
+        try:
+            return self.journal[key]
+        except KeyError:
+            return self.wrapped_db[key]
```

Comment:
I think if we would want to optimize this for performance than we would make the `get(key`) access to return `None` (or a special `Empty` to support `None` as legal value) so that we don't have to loop through all the changesets twice (once for `exists()`, again for `get` if it exists). And then if the returned value is `None` / `Empty` we would look in the underlying db. Downside is that this handling of non-existing key access would be a bit non-standard.

Reply:
Sounds right. Our delete mechanism already assumes that `None` is not an allowed value, it seems.

Reply:
@pipermerriam @carver let me admit that this benchmark is very biased as we are going to look up a key 1 million times in the `JournalDB` that only exists in the underlying db and each time we are going through 10 changesets.

But it shows that not doing the exception ping pong is 5.5 times faster

```python
    memory_db.set(b'x', b'x')
    for _ in range(10):
        journal_db.record()

    then = time.time()
    for _ in range(1000000):
        journal_db.get(b'x')
    now = time.time()

    duration = now - then
```

With the current version this takes about 8.5 seconds to run vs 1.5 seconds in the optimized not-exception based version that I haven't pushed yet.

Reply:
I have to back off I'm afraid. My initial test implementation was flawed and the proper implementation that passes all tests is only 1.5 x faster. Not sure if it's worth it. I'll push it for discussion.